[SPARK-17173][SPARKR] R MLlib refactor, cleanup, reformat, fix deprecation in test

felixcheung opened the pull request:
## What changes were proposed in this pull request?

refactor, cleanup, reformat, fix deprecation in test
## How was this patch tested?

unit tests, manual tests

felixcheung committed:
refactor, cleanup, fix deprecation in test

SparkQA commented:
**[Test build #64150 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64150/consoleFull)** for PR 14735 at commit [`d73a2c0`](https://github.com/apache/spark/commit/d73a2c04ef3ba0cc033f79d4c83e8dd659f71b47).

felixcheung commented:
@junyangq @mengxr @yanboliang @shivaram

felixcheung committed:
add print.summary.GeneralizedLinearRegressionModel test

SparkQA commented:
**[Test build #64151 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64151/consoleFull)** for PR 14735 at commit [`3ea30bb`](https://github.com/apache/spark/commit/3ea30bb3b5d22626f6de6e0699504180f267dfdc).

SparkQA commented:
**[Test build #64150 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64150/consoleFull)** for PR 14735 at commit [`d73a2c0`](https://github.com/apache/spark/commit/d73a2c04ef3ba0cc033f79d4c83e8dd659f71b47).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64150/
Test FAILed.

felixcheung committed:
fix bugs in test

SparkQA commented:
**[Test build #64154 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64154/consoleFull)** for PR 14735 at commit [`1ef18d6`](https://github.com/apache/spark/commit/1ef18d6abfe854c95e0323a406065d9ee4f11c15).

SparkQA commented:
**[Test build #64151 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64151/consoleFull)** for PR 14735 at commit [`3ea30bb`](https://github.com/apache/spark/commit/3ea30bb3b5d22626f6de6e0699504180f267dfdc).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64151/
Test FAILed.

SparkQA commented:
**[Test build #64154 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64154/consoleFull)** for PR 14735 at commit [`1ef18d6`](https://github.com/apache/spark/commit/1ef18d6abfe854c95e0323a406065d9ee4f11c15).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64154/
Test FAILed.

shivaram commented:
This seems a big enough change that it might be good to have a JIRA for this ?

felixcheung committed:
fix test

SparkQA commented:
**[Test build #64157 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64157/consoleFull)** for PR 14735 at commit [`30815e0`](https://github.com/apache/spark/commit/30815e067a37175e0f5d4539c80db6b0ec6cc159).

felixcheung commented:
This also tighten the signature for mllib by removing the previously unused `...`:
```
"summary", signature(object = "GeneralizedLinearRegressionModel")
print.summary.GeneralizedLinearRegressionModel
"summary", signature(object = "NaiveBayesModel")
"summary", signature(object = "IsotonicRegressionModel")
"fitted", signature(object = "KMeansModel")
"summary", signature(object = "KMeansModel")
"spark.naiveBayes", signature(data = "SparkDataFrame", formula = "formula"
"summary", signature(object = "GaussianMixtureModel")
```

SparkQA commented:
**[Test build #64157 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64157/consoleFull)** for PR 14735 at commit [`30815e0`](https://github.com/apache/spark/commit/30815e067a37175e0f5d4539c80db6b0ec6cc159).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64157/
Test PASSed.

felixcheung committed:
revert signature change causing check cran warning

* checking S3 generic/method consistency ... WARNING
print:
  function(x, ...)
print.summary.GeneralizedLinearRegressionModel:
  function(x)

print:
  function(x, ...)
print.summary.GeneralizedLinearRegressionModel:
  function(x)

SparkQA commented:
**[Test build #64161 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64161/consoleFull)** for PR 14735 at commit [`b42e10c`](https://github.com/apache/spark/commit/b42e10c84f122774411e6d9bdc08b27951161272).

SparkQA commented:
**[Test build #64161 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64161/consoleFull)** for PR 14735 at commit [`b42e10c`](https://github.com/apache/spark/commit/b42e10c84f122774411e6d9bdc08b27951161272).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64161/
Test PASSed.

mengxr commented:
@junyangq Could you help review this PR?

junyangq commented on R/pkg/R/mllib.R:
```diff
@@ -610,8 +616,8 @@ setMethod("summary", signature(object = "KMeansModel"),
-            return(list(coefficients = coefficients, size = size,
-                   cluster = cluster, is.loaded = is.loaded))
+            list(coefficients = coefficients, size = size,
+                   cluster = cluster, is.loaded = is.loaded)
```
align list arguments?

junyangq commented on R/pkg/R/mllib.R:
```diff
@@ -1052,8 +1034,8 @@ setMethod("summary", signature(object = "GaussianMixtureModel"),
-            return(list(lambda = lambda, mu = mu, sigma = sigma,
-                   posterior = posterior, is.loaded = is.loaded))
+            list(lambda = lambda, mu = mu, sigma = sigma,
+                   posterior = posterior, is.loaded = is.loaded)
```
alignment?

junyangq commented on R/pkg/R/mllib.R:
```diff
@@ -1163,17 +1145,17 @@ setMethod("spark.als", signature(data = "SparkDataFrame"),
+            itemFactors <- dataFrame(callJMethod(jobj, "itemFactors"))
+            rank <- callJMethod(jobj, "rank")
+            list(user = user, item = item, rating = rating, userFactors = userFactors,
+                          itemFactors = itemFactors, rank = rank)
```
same as above..

junyangq commented on R/pkg/inst/tests/testthat/test_mllib.R:
```diff
@@ -95,6 +95,10 @@ test_that("spark.glm summary", {
 
+  out <- capture.output(print(stats))
+  expect_match(out[2], "Deviance Residuals:")
+  expect_true(any(grepl("AIC: 59.22", out)))
```
Are we trying to test `print.summary.GeneralizedLinearRegressionModel` function here?

yanboliang commented on R/pkg/R/mllib.R:
```diff
@@ -1027,7 +1009,7 @@ setMethod("spark.gaussianMixture", signature(data = "SparkDataFrame", formula =
 #' @export
 #' @note summary(GaussianMixtureModel) since 2.1.0
 setMethod("summary", signature(object = "GaussianMixtureModel"),
-          function(object, ...) {
```
Then annotation ```@param ... currently not used argument(s) passed to the method.``` is not used?

felixcheung replied:
It's in the generic defined in a different file generics.R
We have discussed this at length, this is unfortunately a compromise we need to take at this point. We are going to see some documentation tags spread a bit from now on.

Long story short, CRAN checks warn about undocumented parameter; we need to have the generic with `...` to match the base::summary signature; we shouldn't add `...` in the actual definition since we don't use it and having it weakens the function contract (which is discouraged).

yanboliang replied:
Thanks for clarification.

yanboliang commented on R/pkg/R/mllib.R:
```diff
@@ -499,11 +505,11 @@ setMethod("predict", signature(object = "IsotonicRegressionModel"),
 #' @export
 #' @note summary(IsotonicRegressionModel) since 2.1.0
 setMethod("summary", signature(object = "IsotonicRegressionModel"),
-          function(object, ...) {
```
The annotation ```@param ... Other optional arguments to summary of an IsotonicRegressionModel``` is not be used and should be removed?

felixcheung replied:
ditto.

felixcheung committed:
fix alignment

SparkQA commented:
**[Test build #64198 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64198/consoleFull)** for PR 14735 at commit [`d727093`](https://github.com/apache/spark/commit/d727093092514f4e49f0a2f2e0814ef5df4c71ed).

SparkQA commented:
**[Test build #64198 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64198/consoleFull)** for PR 14735 at commit [`d727093`](https://github.com/apache/spark/commit/d727093092514f4e49f0a2f2e0814ef5df4c71ed).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64198/
Test PASSed.

felixcheung commented:
any more thought? I'd like to merge to facilitate more R mllib work.

junyangq commented:
LGTM

shivaram commented:
@felixcheung I didn't look at the code very closely, but will this change be required in `branch-2.0` as well ? If so the merge might be hard to

felixcheung commented:
I don't think they should be required for branch 2.0 - some part of the signature change with ... is likely good to have for consistency but those might also be "breaking" for a *.0.1 release

If we think we should - since we did make some changes like that in 2.0.0 branch - I could open a PR for the branch separately.

shivaram commented:
Leaving it out of branch-2.0 sounds good to me.
